lib/InitModelTypes/InitModelTypes.cpp

```diff
@@ -649,108 +569,106 @@ initModelTypesImpl(FunctionMetadataCache &Cache,
     }
   } break;
 
   default:
     break;
   }
 
-  rc_return Type;
+  // We didn't manage to find a suitable type: fall back to the LLVM one.
+  rc_return std::nullopt;
 }
 
 static RecursiveCoroutine<ModelTypesMap>
 initModelTypesImpl(FunctionMetadataCache &Cache,
                    const llvm::Function &F,
                    const model::Function *ModelF,
-                   const Binary &Model,
+                   const model::Binary &Model,
                    bool PointersOnly,
                    llvm::SmallPtrSet<const llvm::PHINode *, 8>
                      VisitedPHIs = {}) {
 
   ModelTypesMap TypeMap;
 
-  const model::Type *Prototype = ModelF->prototype(Model).getConst();
-  revng_assert(Prototype);
-
-  addArgumentsTypes(F, Prototype, Model, TypeMap, PointersOnly);
+  const auto *Prototype = Model.prototypeOrDefault(ModelF->prototype());
+  auto Layout = abi::FunctionType::Layout::make(*Prototype);
+  addArgumentsTypes(F, Layout, Model, TypeMap, PointersOnly);
 
   for (const BasicBlock *BB : RPOT<const llvm::Function *>(&F)) {
     for (const Instruction &I : *BB) {
-      std::optional<QualifiedType> Type = initModelTypesImpl(Cache,
-                                                             I,
-                                                             F,
-                                                             ModelF,
-                                                             Model,
-                                                             PointersOnly,
-                                                             TypeMap,
-                                                             VisitedPHIs);
+      std::optional<model::UpcastableType> Result = rc_recur
+        initModelTypesImpl(Cache,
+                           I,
+                           F,
+                           ModelF,
+                           Model,
+                           PointersOnly,
+                           TypeMap,
+                           VisitedPHIs);
       if (PointersOnly) {
         // Skip if it's not a pointer and we are only interested in pointers
-        if (Type and Type->isPointer())
-          TypeMap.insert({ &I, *Type });
+        if (Result.has_value() and !Result->empty() and (*Result)->isPointer())
+          TypeMap.insert({ &I, std::move(*Result) });
 
-      } else {
-        // As a fallback, use the LLVM type to build the QualifiedType
-        if (not Type and I.getType()->isIntOrPtrTy())
-          Type = llvmIntToModelType(I.getType(), Model);
+      } else if (Result.has_value()) {
+        TypeMap.insert({ &I, std::move(*Result) });
 
-        if (Type)
-          TypeMap.insert({ &I, *Type });
+      } else if (I.getType()->isIntOrPtrTy()) {
+        // As a fallback, use the LLVM type
+        TypeMap.insert({ &I, llvmIntToModelType(I.getType(), Model) });
+
+      } else if (auto *Call = llvm::dyn_cast<llvm::CallInst>(&I)) {
+        // TODO: is there more we can check here?
+
```

Comment:
> if this `else if` is taken we're not calling `revng_abort`, even if we should

Ah, right!

That's the reason for the comment. I was triggering the assert. If you look at the old version, it was never asserting on anything. so my instinct was to catch all those cases. And practically, I only ever saw it trigger on calls - hence the TODO to investigate it more thoroughly.

Reply:
Just drop it please.

Reply:
Drop what? The comment? Because without this if it asserts. And an if with no body looks way too weird.

Reply:
I have to look into the specifc cases where it asserts. Do you have a reproducer? If not you can set it up on `macsux` in my orchestra

Reply:
Your orchestra is occupied. I'll use mine instead (will do as soon as I push the branch).

Reply:
Thanks. I kinda lost track of which orchestras are occupied and which aren't.

Reply:
Ok, this change totally makes sense. The only situation where we expect this to fail is for `llvm::Instruction` whose type is not `void`, an integer, nor a pointer. And in the IR generated by `revng` this can only happen with function calls.